docs/tutorial/source-plugin-tutorial.md

```diff
@@ -0,0 +1,984 @@
+      nodes {
+        id
+        slug
+        description
+        imgAlt
+        author {
+          id
+          name
+        }
+        remoteImage {
+          id
+          childImageSharp {
+            id
+            fluid {
+              ...GatsbyImageSharpFluid
+            }
+          }
+        }
+      }
+    }
+  }
+`
+```
+
+This code uses a [page query](/docs/page-query/) to fetch all posts and provide them to the component in the `data` prop at build time. The posts are looped through and rendered to the DOM.
+
+### Using plugin options to customize plugin usage
+
+You can pass options into a plugin through the `gatsby-config.js`. Update the code where your plugin is installed in the `example-site`, changing it from a string, to an object with a `resolve` and `options` key.
+
+```javascript:title=example-site/gatsby-config.js
+module.exports = {
+  plugins: [
+    // highlight-start
+    {
+      resolve: require.resolve(`../source-plugin`),
+      options: {
+        previewMode: true,
+      },
+    },
+    // highlight-end
+    `gatsby-plugin-sharp`,
+    `gatsby-transformer-sharp`,
+  ],
+}
+```
+
+Now you can access the second argument on Node APIs to access the options passed in. Add an argument called `pluginOptions` to your `sourceNodes` function.
+
+```javascript:title=source-plugin/gatsby-node.js
+exports.sourceNodes = async (
+  { actions, createContentDigest, createNodeId, getNodesByType },
+  pluginOptions // highlight-line
+) => {
+  const { createNode, touchNode, deleteNode } = actions
+  // highlight-start
+  console.log(pluginOptions.previewMode) // true
+  // highlight-end
+
+  // ... other code removed for brevity
+}
+```
+
+Options can be a good way of providing conditional paths to logic that you as a plugin author want to provide or limit.
+
+### Proactively updating data with subscriptions
+
+The data sourced for your site was fetched using Apollo Client which supports subscriptions. GraphQL subscriptions _listen_ for changes in data and return changes to the GraphQL client. Your source plugin is able to listen—or subscribe—to the new data that is incoming. That means if a post has something on it updated, your source plugin can listen for that change and update the data in your site without having to restart your site, neat!
+
+The API you connect to needs to provides support for live changes to data in order for this to be possible. You can read about other options for live data updates in the [creating a source plugin guide](/docs/creating-a-source-plugin/).
+
+You already setup your client to handle subscriptions by providing a websocket link (`ws://localhost:4000` or `ws://gatsby-source-plugin-api.glitch.me/`). Now you need to add some logic to your `sourceNodes` function to handle updating and deleting nodes, rather than just creating them. The first step will be touching nodes, to make sure that Gatsby doesn't discard the nodes that don't get updated when `sourceNodes` gets called.
```

Comment:
```suggestion
You already set up your client to handle subscriptions by providing a websocket link (`ws://localhost:4000` or `ws://gatsby-source-plugin-api.glitch.me/`). Now you need to add some logic to your `sourceNodes` function to handle updating and deleting nodes, rather than just creating them. The first step will be touching nodes, to make sure that Gatsby doesn't discard the nodes that don't get updated when `sourceNodes` gets called.
```